core/src/main/java/de/bwaldvogel/mongo/MongoServer.java

```diff
@@ -48,7 +49,15 @@
     private SslContext sslContext;
 
     public MongoServer(MongoBackend backend) {
-        this.backend = backend;
+        this(backend, false);
+    }
+
+    public MongoServer(MongoBackend backend, boolean oplogEnabled) {
```

Comment:
I don't think adding constructor argument is better than manual composition of server(s) (creating new `OplogBackend`)

### Example
```java
server  = new OplogServer(server); // this version is preferred 
// vs 
server = new MongoServer(backend, true); // what is true argument exactly ? 
```

Reply:
I’m not really a fan of the boolean constructor parameter either.
At the moment I would keep the oplog disabled by the default but it should be easy to enable it.
I was thinking about:

```java
server = new MongoServer(backend);
server.enableOplog();
```
or:
```java
backend.enableOplog();
server = new MongoServer(backend);
```
Maybe even both and `server.enableOplog();` would just delegate to `backend.enableOplog()`. (…and of course there should be also `disableOplog()`).

I like the idea to implement as much oplog functionality using composition instead of inheritance. However, I think some oplog related code should go to `AbstractMongoBackend` directly instead of a subclass.

What do you think about adding an `Oplog` interface that basically defines the CRUD operations?
Every `AbstractMongoBackend` would have a field (never null!) that references an instance of `Oplog`. The default implementation of `Oplog` would be an instance of `NoopOplog` which doesn’t do anything. When someone tries to query the oplog from `NoopOplog` we could even throw an `MongoServerException` with a user friendly message explaining how to enable the oplog.
If someone enables the oplog, we would replace the `NoopOplog` instance by something like `CollectionBackedOplog` that would use an instance of `MongoCollection` to store/fetch the oplog.
Keep in mind that different backends could decide not to manage the oplog in its own collection but retrieve it from the database itself. Maybe something like that would be possible in the Postgres backend, i.e. retrieving the oplog from the transaction log of Postgres.
I’m not saying that this should be implemented in the first iteration but I would like to keep it in mind when choosing an implementation approach.

Reply:
That sound good to me.

Reply:
Do we really need to enable / disable oplog at runtime ? 
Why not have this as constructor argument to eg. `AbstractMongoBackend` ?

Reply:
I agree with @asereda-gs I think is better to enable or disable the oplog when the server is created and not change it through out the life of the instance.

Reply:
By constructor argument I meant `Oplog` instance (not `true` / `false`).

IE set `Clock` / `Oplog` dependencies at backend / server instantiation.

Reply:
IMO we shouldn’t let the user provide an instance of `Oplog`. The API would be to complicated, we would expose implementation details and we couldn’t easily rename/refactor the oplog.

What do you think about something like `MongoServerOptions` that could be provided to the `MongoServer` constructor next to the backend?

Reply:
> What do you think about something like MongoServerOptions

Yes that's more or less how I was thinking. Have `config / setup / options` class and a factory method to create the backend:

```java
class MongoBackend {
   
   static MongoBackend create() {
     return create(MongoOptions.defaultOptions);
   }

   static MongoBackend create(MongoOptions options) {
     // possible decorated mongo instance
     // or use default implementation
   }
}
``` 

Another option is use a builder for MongoBackend.